IOS/SwiftSeedProject/SwiftSeedProject/Infrastructure/CoreDataStack/CoreDataStack.swift

```diff
@@ -0,0 +1,118 @@
+        let fileManager = FileManager.default
+        let documentsDirectoryURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
+        return documentsDirectoryURL.appendingPathComponent(storeName) as NSURL
+    }
+    
+    private lazy var managedObjectModel: NSManagedObjectModel? = {
+        // The mom and momd files are compiled versions of xcdatamodel and xcdatamodeld files.
+        guard let modelURL = Bundle.main.url(forResource: self.modelName, withExtension: "momd") else {
+            return nil
+        }
+        return NSManagedObjectModel(contentsOf: modelURL)
+    }()
+    
+    private lazy var persistentStoreCoordinator: NSPersistentStoreCoordinator? = {
+        guard let managedObjectModel = self.managedObjectModel else {
+            return nil
+        }
+        return NSPersistentStoreCoordinator(managedObjectModel: managedObjectModel)
+    }()
+    
+    private lazy var privateManagedObjectContext: NSManagedObjectContext = {
+        let managedObjectContext = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
+        managedObjectContext.persistentStoreCoordinator = self.persistentStoreCoordinator
+        return managedObjectContext
+    }()
+    
+    public private(set) lazy var mainManagedObjectContext: NSManagedObjectContext = {
+        let managedObjectContext = NSManagedObjectContext(concurrencyType: .mainQueueConcurrencyType)
+        managedObjectContext.parent = self.privateManagedObjectContext
+        return managedObjectContext
+    }()
+    
+    init(modelName: String, completion: @escaping CoreDataStackCompletion) {
+        self.modelName = modelName
+        self.completion = completion
+        setupCoreDataStack()
+    }
+    
+    private func setupCoreDataStack() {
+        let _ = mainManagedObjectContext.persistentStoreCoordinator
+        
+        DispatchQueue.global().async {
+            self.addPersistentStore()
+            DispatchQueue.main.async { self.completion() }
+        }
+    }
+    
+    private func addPersistentStore() {
+        guard let persistentStoreCoordinator = persistentStoreCoordinator else { fatalError("Unable to Initialize Persistent Store Coordinator") }
+        
+        let persistentStoreURL = self.persistentStoreURL
+        do {
+            let options = [ NSMigratePersistentStoresAutomaticallyOption : true, NSInferMappingModelAutomaticallyOption : true ]
+            try persistentStoreCoordinator.addPersistentStore(ofType: NSSQLiteStoreType, configurationName: nil, at: persistentStoreURL as URL, options: options)
+            
+        } catch {
+            let addPersistentStoreError = error as NSError
+            
+            print("Unable to Add Persistent Store")
+            print("\(addPersistentStoreError.localizedDescription)")
+        }
+    }
+    
+    // MARK: - Helper Methods
+    
+    public func saveChanges() {
+        mainManagedObjectContext.performAndWait({
```

Comment:
Have you heard of DRY principle? 😄 Can you try avoid repeating almost the same code twice inside this method?